packages/compiler/lib/decorators.ts

```diff
@@ -621,3 +621,25 @@ export function isKey(program: Program, property: ModelTypeProperty) {
 export function getKeyName(program: Program, property: ModelTypeProperty): string {
   return program.stateMap(keyKey).get(property);
 }
+
+export function $rename({ program }: DecoratorContext, entity: Type, newName: Type) {
+  if (newName.kind === "TemplateParameter") {
+    // Don't execute when this decorator is being applied within a templated type
+    return;
+  }
+
+  if (
+    !validateDecoratorTarget(program, entity, "@rename", ["Model", "ModelProperty", "Operation"])
+  ) {
+    // TODO: Diagnostic
```

Comment:
Whoops, forgot to add diagnostics!  Will add in a follow-up commit.

Reply:
it should emit the diagnostic here for you with the validate

Reply:
Seems like just remove the TODO here.  We *should* have test coverage to verify the diagnostic